Contest_API.md

```diff
@@ -605,29 +600,28 @@ Provides information on the current contest.
 
 The following endpoints are associated with contest:
 
-| Endpoint         | Mime-type        | Required? | Description                                                                 
-| :--------------- | :--------------- | :-------- | :----------
-| `/contests`      | application/json | yes       | JSON array of all contests with properties as defined in the table below.
```

Comment:
I'm not so convinced that there are many endpoints that *must* always be present, if by must we mean that you could not have any useful service without them.

These would be my candidates:

- `/` because it tells you which version of the API you have.
- `/contests/<id>/access` because it tells you what other endpoints are availeble.
- `/contests` because otherwise you couldn't know where to get the endpoint above.

The argument for why e.g. `/teams` and `/problems` wouldn't be *always* *strictly* required is that you could use an API for partial configuration of ome other service, and maybe that configuration source doesn't have the teams and problems? OTOH, the counterarguemnt would be that in that case you could just let `/problems` and `/teams` return empty results, so it would not be any additional burden to require them.

In summary, I'm just not sure.

Reply:
I'd say if you have `/contests` it would be very strange to not also have `/contests/<id>`.

And I kinda agree that there *could* be very strange cases where you don't have `/problems`. I'm not sure if it ever happens in practice and if we are not making it too complex for ourselves by making everything optional though.

Reply:
I don't love it, but maybe @niemela is right and we should look at this like the recent change to the CAF, where instead of having a bunch of required files we limit this and look at it as 'useful subsets' for certain scenarios.

I think I'd prefer that / and /contests always exist, and when a particular contest exists then /access, /scoreboard (even a dummy/empty response), and /event-feed also exist.

Reply:
> I'd say if you have `/contests` it would be very strange to not also have `/contests/<id>`.

Yes, that was meant to be implied.

> 'm not sure if it ever happens in practice 

Agreed.

> and if we are not making it too complex for ourselves by making everything optional though.

But, either it never happens, and then it's really not complex. We are not *adding* anything to make things optional, or if it *does* happen, then at least it  wasn't unnecessary, and it's still not very complex?

> I think I'd prefer that / and /contests always exist, and when a particular contest exists then /access, /scoreboard (even a dummy/empty response), and /event-feed also exist.

But scoreboard and event-feed would be two of the things that would be among the more likely/useful to leave out. An API with everything except scoreboard and/or event-feed would certainly be useful, and it would also be *a lot* easier to implement.

Reply:
Hmm. I guess I grudgingly agree? :)

I'll fall back to / and /contests always existing, and /access exists the moment the contest does.

Access endpoints seems a bit weird for /scoreboard and /event-feed though, maybe these should be capabilities?? We should avoid doing everything in this PR though, I would be fine closing on properties and the basics of which endpoints are mandatory and then handling things like this via another PR or discussion on Monday.

Reply:
> I'd really like to see a "minimal" set of required endpoints and attributes

What minimum set of endpoints do you suggest?

Note that it's not actually a big ask to require an endpoint. All endpoints (except access and state, and access is already required anyway) are sets, which means that the empty set *is* an acceptable implementation. E.g. if awards was required you could trivially implement awards by always returning `[]`, i.e. you would "support" awards, you would just never have any.

Reply:
I don't think there's any point in requiring endpoints where the expectation is you might always return `[]`. Even for awards, there is no point supporting it unless at a minimum some other tool could PUT and you return that set - i.e. a CCS that can't assign awards but could still be the single source of truth.

From a spec perspective, I think I'm still with my previous comment - the required set should be very small, but instead maybe we define some SHOULDs or logical sets as the CAF examples or CCS spec is doing? e.g. for configuration the CMS (or we) could host a live API with just the registration data; for finals the CCS needs a bigger set?

Reply:
> I don't think there's any point in requiring endpoints where the expectation is you might always return `[]`.

Right, but that's true for pretty much *all* endpoints. Which kinda was my point?

Although the point does cut both ways, i.e.:
- It's no big hurdle to be forced to implement (say) clarifications, since the server could just not have any and always return `[]`. Thus there is no reason to *not* require it.
- There is no practical difference from getting a 404 from a non-implemented endpoint compared to getting a `[]` for a unused (but technically implemented) endpoint. Except in the latter case `/access` would still list the unused endpoint and would this be slightly less useful. Thus there is no reason *to* require it.

...

Actually after writing down the argument in either direction I'm leaning a but for towards *not* requiring more endpoints.

Reply:
Ok, so I'd propose the following as a "minimal" useful set of endpoints and properties a server should implement:
```json
{
   "capabilities": [],
   "endpoints": [
      {
         "type": "contest",
         "properties": [
            "id", "name", "start_time", "duration"
         ]
      },
      {
         "type": "judgement-types",
         "properties": [
            "id", "name", "solved"
         ]
      },
      {
         "type": "languages",
         "properties": [
            "id", "name"
         ]
      },
      {
         "type": "problems",
         "properties": [
            "id", "name"
         ]
      },
      {
         "type": "teams",
         "properties": [
            "id", "name"
         ]
      },
      {
         "type": "submissions",
         "properties": [
            "id", "language_id", "problem_id", "team_id", "time"
         ]
      },
      {
         "type": "judgements",
         "properties": [
            "id", "submission_id", "judgement_type_id"
         ]
      }
   ]
}
```
I think that for any reasonably normal contest, this is pretty much the bare minimum that allows a client to have the basic information do some generic useful things.

Reply:
I think Fredrik's point (a few ago...) was what if (e.g.) someone wanted to host the CMS data on a server instead of a file system? With the list above this would be invalid, or they'd have to fake some empty judgement types, languages, and submissions.

Like the CAF, I am wondering if the API spec really needs all of these required, or if your list above is more of a minimal CCS implementation of the spec.

Reply:
> ```json
>          "type": "contest",
>          "properties": [
>             "id", "name", "start_time", "duration"
> ```

If we don't include `penalty_time` it needs a default value. (0 would be natural).

> ```json
>          "type": "languages",
> ```

I don't think `languages` is needed. Specifically in a "submit the output instead of the code" kind of contest.

> ```json
>          "type": "submissions",
>          "properties": [
>             "id", "language_id", "problem_id", "team_id", "time"
> ```

If we drop `languages` above we must also drop `language_id` here.

I agree that this a a good very minimal contest setup. See @deboer-tim post above though (which I agree with)

Reply:
Ok, but in response to @deboer-tim's comment: I think of this explicitly as a "should" implement for CCSs, and don't consider this a requirement e.g. for a contest archive.

And I agree that language could indeed be dropped. Although I'd say that it would almost always be available, I think it is not _that_ useful for a typical client, so leaving languages out is ok to me.

Indeed, `penalty_time = 0` seems a reasonable default if nothing is included.

Reply:
I added a default for `penalty_time`.

Reply:
Had a chat with Fredrik yesterday, and I'm more comfortable with not specifying required properties in each endpoint. I do think we need to document this somewhere for different uses though. The CAF does this in the examples section; now we do it for the API in the CCS API, but don't cover the additional CCS requirements for finals or other cases. I think we should go ahead and merge this soon (it's now blocking other cases), and talk on the next call about how to cover this.